File: oa/api/role/role.go
package role

import (
	"github.com/google/uuid"
	"github.com/veypi/OneBD/rest"
	"oa/cfg"
	M "oa/models"
	"strings"
)

func useRole(r rest.Router) {
	r.Get("/:role_id", roleGet)
	r.Get("/", roleList)
	r.Patch("/:role_id", rolePatch)
	r.Post("/", rolePost)
	r.Delete("/:role_id", roleDelete)
}
func roleGet(x *rest.X) (any, error) {
	opts := &M.RoleGet{}
	err := x.Parse(opts)
	if err != nil {
		return nil, err
	}
	data := &M.Role{}

	err = cfg.DB().Where("id = ?", opts.ID).First(data).Error

	return data, err
}
func roleList(x *rest.X) (any, error) {
	opts := &M.RoleList{}
	err := x.Parse(opts)
	if err != nil {
		return nil, err
	}
	data := make([]*M.Role, 0, 10)

	query := cfg.DB()
	if opts.Name != nil {
		query = query.Where("name LIKE ?", opts.Name)
	}
	err = query.Find(&data).Error

	return data, err
}
func rolePatch(x *rest.X) (any, error) {
	opts := &M.RolePatch{}
	err := x.Parse(opts)
	if err != nil {
		return nil, err
	}
	data := &M.Role{}

	err = cfg.DB().Where("id = ?", opts.ID).First(data).Error
	if err != nil {
		return nil, err
	}
	optsMap := make(map[string]interface{})
	if opts.Name != nil {
		optsMap["name"] = opts.Name
	}
	if opts.Des != nil {
		optsMap["des"] = opts.Des
	}
	if opts.AppID != nil {
		optsMap["app_id"] = opts.AppID
	}
	err = cfg.DB().Model(data).Updates(optsMap).Error

	return data, err
}
func rolePost(x *rest.X) (any, error) {
	opts := &M.RolePost{}
	err := x.Parse(opts)
	if err != nil {
		return nil, err
	}
	data := &M.Role{}

	data.ID = strings.ReplaceAll(uuid.New().String(), "-", "")
	data.Name = opts.Name
	data.Des = opts.Des
	data.AppID = opts.AppID
	err = cfg.DB().Create(data).Error

	return data, err
}
func roleDelete(x *rest.X) (any, error) {
	opts := &M.RoleDelete{}
	err := x.Parse(opts)
	if err != nil {
		return nil, err
	}
	data := &M.Role{}

	err = cfg.DB().Where("id = ?", opts.ID).Delete(data).Error

	return data, err
}

File: oaweb/composables/api/role.ts
//
// Copyright (C) 2024 veypi <i@veypi.com>
// 2024-10-11 00:18:06
// Distributed under terms of the MIT license.
//

import webapi from "./webapi"
import * as models from "./models"
export interface ListOpts { 
  name?: string
}
export function List(json: ListOpts) {
  return webapi.Get<models.Role>(`/role`, { json })
}

export function Get(role_id: string) {
  return webapi.Get<models.Role>(`/role/${role_id}`, {  })
}

export interface PatchOpts { 
  name?: string
  des?: string
  app_id?: string
}
export function Patch(role_id: string, json: PatchOpts) {
  return webapi.Patch<models.Role>(`/role/${role_id}`, { json })
}

export function Delete(role_id: string) {
  return webapi.Delete<models.Role>(`/role/${role_id}`, {  })
}

export interface PostOpts { 
  name: string
  des: string
  app_id: string
}
export function Post(json: PostOpts) {
  return webapi.Post<models.Role>(`/role`, { json })
}

